File: lua/CSharp/LuaLexerBase.cs
using System;
using System.CodeDom;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.IO;
using Antlr4.Runtime;
using Microsoft.Build.Utilities;

public abstract class LuaLexerBase : Lexer
{
    private int start_line;
    private int start_col;

    public LuaLexerBase(ICharStream input)
            : base(input)
    {
    }

    public LuaLexerBase(ICharStream input, TextWriter output, TextWriter errorOutput)
            : base(input, output, errorOutput)
    {
    }

    public void HandleComment()
    {
        start_line = this.Line;
        start_col = this.Column - 2;
        var cs = (ICharStream)InputStream;
        if (cs.LA(1) == '[')
        {
            int sep = skip_sep(cs);
            if (sep >= 2)
            {
                read_long_string(cs, sep);
                return;
            }
        }
        while (cs.LA(1) != '\n' && cs.LA(1) != -1)
        {
            cs.Consume();
        }
    }

    private void read_long_string(ICharStream cs, int sep)
    {
        bool done = false;
        cs.Consume();
        for (; ; )
        {
            var c = cs.LA(1);
            var cc = (char)c;
            switch (c)
            {
                case -1:
                    done = true;
                    IAntlrErrorListener<int> listener = ErrorListenerDispatch;
                    listener.SyntaxError(ErrorOutput, this, 0, this.start_line, this.start_col, "unfinished long comment", null);
                    break;
                case ']':
                    if (skip_sep(cs) == sep)
                    {
                        cs.Consume();
                        done = true;
                    }
                    break;
                default:
                    if (cs.LA(1) == -1)
                    {
                        done = true;
                        break;
                    }
                    cs.Consume();
                    break;
            }
            if (done) break;
        }
    }

    private int skip_sep(ICharStream cs)
    {
        int count = 0;
        int s = cs.LA(1);
        char ss = (char)s;
        cs.Consume();
        while (cs.LA(1) == '=')
        {
            cs.Consume();
            count++;
        }
        if (cs.LA(1) == s) count += 2;
        else if (count == 0) count = 1;
        else count = 0;
        return count;
    }

    public bool IsLine1Col0()
    {
        var cs = (ICharStream)InputStream;
        if (cs.Index == 1) return true;
        return false;
    }
}

File: lua/Dart/LuaLexerBase.dart
import 'package:antlr4/antlr4.dart';
import 'dart:io';
import 'dart:convert';

abstract class LuaLexerBase extends Lexer
{
    int start_line = 0;
    int start_col = 0;

    LuaLexerBase(CharStream input) : super(input)
    {
    }

    bool IsColumnZero()
    {
        return charPositionInLine == 0;
    }

    void HandleComment()
    {
        start_line = this.line;
        start_col = this.charPositionInLine - 2;
        var cs = this.inputStream;
        var c = cs.LA(1);
        var yo = '['.codeUnitAt(0);
        if (cs.LA(1) == '['.codeUnitAt(0))
        {
            int sep = skip_sep(cs);
            if (sep >= 2)
            {
                read_long_string(cs, sep);
                return;
            }
        }
        for (;;) {
            var c2 = cs.LA(1);
            var yo2 = '\n'.codeUnitAt(0);
            if (c2 == yo2) break;
            if (c2 == -1) break;
            cs.consume();
        }
    }

    void read_long_string(CharStream cs, int sep)
    {
        bool done = false;
        cs.consume();
        for (; ; )
        {
            var c = cs.LA(1);
            var cc = c;
            var yo = ']'.codeUnitAt(0);
            if (c == -1)
            {
                done = true;
                final listener = this.errorListenerDispatch;
                listener.syntaxError(this, null, this.start_line, this.start_col, "unfinished long comment", null);
            } else if (c == yo) {
                if (skip_sep(cs) == sep)
                {
                    cs.consume();
                    done = true;
                }
            } else {
                if (cs.LA(1) == -1)
                {
                    done = true;
                } else {
                    cs.consume();
                }
            }
            if (done) break;
        }
    }

    int skip_sep(CharStream cs)
    {
        int count = 0;
        var s = cs.LA(1);
        cs.consume();
        var yoeq = '='.codeUnitAt(0);
        while (cs.LA(1) == yoeq)
        {
            cs.consume();
            count++;
        }
        if (cs.LA(1) == s) count += 2;
        else if (count == 0) count = 1;
        else count = 0;
        return count;
    }

    bool IsLine1Col0()
    {
        var cs = this.inputStream;
        if (cs.index == 1) return true;
        return false;
    }
}
